src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,173 @@
+            }
+            head = newNode;
+            if (tail == null) {
+                tail = head;
+            }
+        } else {
+            Node<T> current = getNode(index);
+            newNode.next = current;
+            newNode.prev = current.prev;
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Unable to get element from the list: "
+                + "no such index in list");
+        }
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Unable to set element to the list:"
+                + " no such index in list");
+        }
+        Node<T> current = getNode(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Unable to remove element from the list: "
+                + "no such index in list");
+        }
+
+        Node<T> current = getNode(index);
+
+        if (current.prev == null) {
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+        }
+
+        if (current.next == null) {
+            tail = current.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
+
+        size--;
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+
+        while (current != null) {
+            if ((object == null && current.value == null)
+                    || (object != null && object.equals(current.value))) {
+                if (current.prev == null) {
+                    head = current.next;
+                } else {
+                    current.prev.next = current.next;
+                }
+
+                if (current.next == null) {
+                    tail = current.prev;
+                } else {
+                    current.next.prev = current.prev;
+                }
+
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
+
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
```

Comment:
You have a method `getNode(int index)` that is used in multiple places, which is good. However, consider creating a `checkIndex` method to verify the index is valid, as the index validation logic is repeated in `add(T value, int index)`, `get(int index)`, `set(T value, int index)`, and `remove(int index)` methods.